topostats/tracing/dnatracing.py

```diff
@@ -677,98 +758,203 @@ def findWrithe(self):
+
+        # Testing with a parabola
+        # x = np.linspace(-2, 2, num=self.n_points)
+        # y = x**2
+        # self.splined_traces[0] = np.column_stack((x, y))
+        # self.mol_is_circular[0] = False
+
         # FIXME : Iterate directly over self.splined_traces.values() or self.splined_traces.items()
-        for dna_num in sorted(self.splined_traces.keys()):  # the number of molecules identified
+        for dna_num in sorted(self.simplified_splined_traces.keys()):  # the number of molecules identified
             # splined_traces is a dictionary, where the keys are the number of the molecule, and the values are a
             # list of coordinates, in a numpy.ndarray
-            # if self.mol_is_circular[dna_num]:
+            length = len(self.simplified_splined_traces[dna_num])
             curve = []
             contour = 0
-            coordinates = np.zeros([2, self.neighbours * 2 + 1])
-            for i, (x, y) in enumerate(self.splined_traces[dna_num]):
-                # Extracts the coordinates for the required number of points and puts them in an array
-                if self.mol_is_circular[dna_num] or (
-                    self.neighbours < i < len(self.splined_traces[dna_num]) - self.neighbours
-                ):
-                    for j in range(self.neighbours * 2 + 1):
-                        coordinates[0][j] = self.splined_traces[dna_num][i - j][0]
-                        coordinates[1][j] = self.splined_traces[dna_num][i - j][1]
-
-                    # Calculates the angles for the tangent lines to the left and the right of the point
-                    theta1 = math.atan(
-                        (coordinates[1][self.neighbours] - coordinates[1][0])
-                        / (coordinates[0][self.neighbours] - coordinates[0][0])
-                    )
-                    theta2 = math.atan(
-                        (coordinates[1][-1] - coordinates[1][self.neighbours])
-                        / (coordinates[0][-1] - coordinates[0][self.neighbours])
-                    )
-
-                    left = coordinates[:, : self.neighbours + 1]
-                    right = coordinates[:, -(self.neighbours + 1) :]
-
-                    xa = np.mean(left[0])
-                    ya = np.mean(left[1])
-
-                    xb = np.mean(right[0])
-                    yb = np.mean(right[1])
-
-                    # Calculates the curvature using the change in angle divided by the distance
-                    dist = math.hypot((xb - xa), (yb - ya))
-                    dist_real = dist * self.pixel_size
-                    curve.append([i, contour, (theta2 - theta1) / dist_real])
+            # coordinates = np.zeros([2, self.neighbours * 2 + 1])
+            # dxmean = np.zeros(length)
+            # dymean = np.zeros(length)
+            # gradients = np.zeros([2, self.neighbours * 2 + 1])
+            if self.mol_is_circular[dna_num]:
+                longlist = np.concatenate(
```

Comment:
Why are the co-ordinates concatenated three times over?

Reply:
Otherwise there will be problems with the points at the end of linear molecules